tests/testthat/test-code_dependency.R

```diff
@@ -0,0 +1,393 @@
+  "get_code returns single lines for code put in {} inside expressions",
+  {
+    q <- new_qenv()
+    q <- eval_code(q, expression({
+      a <- 1
+      b <- 2
+    }))
+
+    testthat::expect_identical(
+      get_code(q, names = "a"),
+      "a <- 1"
+    )
+  }
+)
+
+
+
+# functions -------------------------------------------------------------------------------------------------------
+
+testthat::test_that("get_code ignores occurrence in function definition", {
+  q <- new_qenv()
+  q <- eval_code(q, "b <- 2")
+  q <- eval_code(q, "foo <- function(b) { b <- b + 2 }")
+
+  testthat::expect_identical(
+    get_code(q, names = "foo"),
+    "foo <- function(b) {\n    b <- b + 2\n}"
+  )
+})
+
+
+testthat::test_that("get_code ignores effect of the object which occurs in a function definition", {
+  q <- new_qenv()
+  q <- eval_code(q, "b <- 2")
+  q <- eval_code(q, "foo <- function(b) { b <- b + 2 }")
+
+  testthat::expect_identical(
+    get_code(q, names = "b"),
+    c("b <- 2")
+  )
+})
+
+testthat::test_that("get_code detects occurrence of the function object", {
+  q <- new_qenv()
+  q <- eval_code(q, "a <- 1")
+  q <- eval_code(q, "b <- 2")
+  q <- eval_code(q, "foo <- function(b) { b <- b + 2 }")
+  q <- eval_code(q, "b <- foo(a)")
+
+  testthat::expect_identical(
+    get_code(q, names = "b"),
+    c("a <- 1", "b <- 2", "foo <- function(b) {\n    b <- b + 2\n}", "b <- foo(a)")
+  )
+})
+
+testthat::test_that(
+  "Can't detect occurrence of function definition when a formal is named the same as a function",
+  {
+    testthat::skip("This does not return foo definition YET!")
+    q <- new_qenv()
+    q <- eval_code(q, "x <- 1")
+    q <- eval_code(q, "foo <- function(foo = 1) 'text'")
+    q <- eval_code(q, "a <- foo(x)")
+
+    testthat::expect_identical(
+      get_code(q, names = "a"),
+      c("x <- 1", "foo <- function(foo = 1) 'text'", "a <- foo(x)")
+    )
+  }
+)
+
+# $ ---------------------------------------------------------------------------------------------------------------
+
+
+testthat::test_that("get_code understands $ usage and do not treat rhs of $ as objects (only lhs)", {
+  q <- new_qenv()
+  q <- eval_code(q, "x <- data.frame(a = 1:3)")
+  q <- eval_code(q, "a <- data.frame(y = 1:3)")
+  q <- eval_code(q, "a$x <- a$y")
+  q <- eval_code(q, "a$x <- a$x + 2")
+  q <- eval_code(q, "a$x <- x$a")
+
+  testthat::expect_identical(
+    get_code(q, names = "x"),
+    c("x <- data.frame(a = 1:3)")
+  )
+  testthat::expect_identical(
+    get_code(q, names = "a"),
+    c("x <- data.frame(a = 1:3)", "a <- data.frame(y = 1:3)", "a$x <- a$y", "a$x <- a$x + 2", "a$x <- x$a")
+  )
+})
+
+
+# @ ---------------------------------------------------------------------------------------------------------------
+
+testthat::test_that("get_code understands $ usage and do not treat rhs of $ as objects (only lhs)", {
+  testthat::skip("Due ot the error: cannot add bindings to a locked environment when evaluating qenv code")
```

Comment:
hey @gogonzo @chlebowa have you ever tried using S4 class inside `eval_code`? I'm getting an error saying the `eval_code` environment is a locked environment and we can not change S4 objects inside it lol. Sounds like this locking mechanism disables the usage of S4 in `qenv`

Reply:
:eyes:

Reply:
I'm not getting that error...